lib/restforce/concerns/api.rb

```diff
@@ -361,12 +361,17 @@ def upsert(*args)
       # error returned if the external ID provided matches multiple records (in which
       # case the conflicting IDs can be found by looking at the response on the error)
       def upsert!(sobject, field, attrs)
-        external_id = attrs.
-          fetch(attrs.keys.find { |k, v| k.to_s.downcase == field.to_s.downcase }, nil)
-        attrs_without_field = attrs.
-          reject { |k, v| k.to_s.downcase == field.to_s.downcase }
-        response = api_patch "sobjects/#{sobject}/#{field}/#{URI.encode(external_id)}",
-                             attrs_without_field
+        external_id = attrs.delete(field.to_sym)
```

Comment:
I think I'd rather go for the `#extract_case_insensitive_string_or_symbol_key_from_hash` route - it adds a little indirection, but I'd always (personally) tend to want to add a method with a really clear, expressive name rather than a comment.

It might not be reused for now, but when I read the code, I'll instantly know what we're trying to do.

Otherwise, happy with supporting both strings and symbols :+1: